lib/clientv2/node.go

```diff
@@ -59,6 +63,98 @@ func (r nodes) Update(ctx context.Context, res *apiv2.Node, opts options.SetOpti
 
 // Delete takes name of the Node and deletes it. Returns an error if one occurs.
 func (r nodes) Delete(ctx context.Context, name string, opts options.DeleteOptions) (*apiv2.Node, error) {
+	pname, err := names.WorkloadEndpointIdentifiers{Node: name}.CalculateWorkloadEndpointName(true)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get all weps belonging to the node
+	weps, err := r.client.WorkloadEndpoints().List(ctx, options.ListOptions{
+		Prefix: true,
+		Name: pname,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Collate all IPs across all endpoints, and then release those IPs.
+	ips := []net.IP{}
+	for _, wep := range weps.Items {
+		// The prefix match is unfortunately not a perfect match on the Node (since it is theoretically possible for
+		// another node to match the prefix (e.g. a node name of the format <thisnode>-foobar would also match a prefix
+		// search of the node <thisnode>). Therefore, we will also need to check that the Spec.Node field matches the Node.
+		if wep.Spec.Node == name {
+			for _, ip := range wep.Spec.IPNetworks {
+				ips = append(ips, *net.ParseIP(ip))
+			}
+		}
+	}
+	_, err = r.client.IPAM().ReleaseIPs(context.Background(), ips)
+	if err != nil {
+		return nil, err
+	}
+
+	// Delete the weps.
+	for _, wep := range weps.Items {
+		if wep.Spec.Node == name {
+			_, err = r.client.WorkloadEndpoints().Delete(ctx, wep.Namespace, wep.Name, options.DeleteOptions{})
+			if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
+				return nil, err
+			}
+		}
+	}
+
+	// Remove the node from the IPAM data if it exists.
+	err = r.client.IPAM().RemoveIPAMHost(ctx, name)
+	if err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
```

Comment:
I think this logic is the wrong way round, it should be `!ok`

also, you can combine these two ifs:

```
if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok && err != nil {
}
```